Spigot-Server-Patches/0381-Optimize-redstone-algorithm.patch

```diff
@@ -0,0 +1,1185 @@
+From 6a5104193614fbe321dca74d32646a4ad33595fa Mon Sep 17 00:00:00 2001
+From: egg82 <phantom_zero@ymail.com>
+Date: Thu, 27 Sep 2018 01:43:35 -0600
+Subject: [PATCH] Optimize redstone algorithm
+
+This patch implements theosib's redstone algorithms to completely overhaul the way redstone works.
+The new algorithms should be many times faster than current vanilla ones.
+From the original author's comments, it looks like it shouldn't interfere with any redstone save for very extreme edge-cases.
+
+Surprisingly, not a lot was touched aside from a few obfuscation helpers and BlockRedstoneWire.
+A lot of this code is self-contained in a helper class.
+
+Aside from making the obvious class/function renames and obfhelpers I didn't need to modify much.
+Just added Bukkit's event system and took a few liberties with dead code and comment misspellings.
+
+diff --git a/src/main/java/com/destroystokyo/paper/PaperConfig.java b/src/main/java/com/destroystokyo/paper/PaperConfig.java
+index ce5273384..1035dfd43 100644
+--- a/src/main/java/com/destroystokyo/paper/PaperConfig.java
++++ b/src/main/java/com/destroystokyo/paper/PaperConfig.java
+@@ -388,4 +388,14 @@ public class PaperConfig {
+             log("Async Chunks: Enabled - Chunks will be loaded and generated much faster, without lag.");
+         }
+     }
++
++    public static boolean useNewRedstone = false;
++    private static void useNewRedstone() {
++        useNewRedstone = config.getBoolean("settings.use-new-redstone", false);
++        if (useNewRedstone) {
++            log("Using new redstone algorithm by theosib.");
++        } else {
++            log("Using vanilla redstone algorithm.");
++        }
++    }
+ }
+diff --git a/src/main/java/com/destroystokyo/paper/util/RedstoneWireTurbo.java b/src/main/java/com/destroystokyo/paper/util/RedstoneWireTurbo.java
+new file mode 100644
+index 000000000..a4eb697f9
+--- /dev/null
++++ b/src/main/java/com/destroystokyo/paper/util/RedstoneWireTurbo.java
+@@ -0,0 +1,912 @@
++package com.destroystokyo.paper.util;
++
++import java.util.List;
++import java.util.Map;
++import java.util.concurrent.ThreadLocalRandom;
++
++import org.bukkit.event.block.BlockRedstoneEvent;
++
++import com.google.common.collect.Lists;
++import com.google.common.collect.Maps;
++
++import net.minecraft.server.Block;
++import net.minecraft.server.BlockPosition;
++import net.minecraft.server.BlockRedstoneWire;
++import net.minecraft.server.IBlockData;
++import net.minecraft.server.World;
++
++/**
++ * Used for the faster redstone algorithm.
++ * Original author: theosib
++ * 
++ * Ported class names, updated methods, added events, etc
++ * Original comments are (at least mostly) preserved
++ * 
++ * @author egg82

```

Comment:
I'd drop this.

Reply:
Was mostly following Aikar's previous examples. The author bit is a bit much, I agree.